### ISSUE TITLE
new field type: tags

### ISSUE THREAD

**Issue by tablatronix** [n00dles/DM_Matrix#26]
Do you see any benefit of a datatype of 'tags' ?

**Comment by n00dles** [comment]
Yeah why not, the more the merrier I think. 

Means we won't have to come back to it when someone says they'd love a tags field.. 8)

I was also thinking of adding a Slider selector, Multi Image field and a normal dropdown.

**Comment by tablatronix** [comment]
oh is there isn't one is there.
It' s almost as is you need pseudo datatypes.

It seems there are alot of ui stuff you can do for fields, but they all essentially store themselves as text fields.

slider with clamp values
color picker
(array or ordered list) ( multi image field,tags,dropdown,key value pairs ) sortable ?

**Comment by tablatronix** [comment]
I guess I meant to further describe this, but never did, you know when you think something and assume you said it out loud in full detail.

The problem with continuing to add datatypes is that it will get ridiculously complex.

It is as if you need a uilist map for datatypes to editing types.

As i mentioned above

A slider could be an int, float etc. But the data to store its clamps and multiplier etc to display it would not be in the data itself.

A color selector could store data as an rgb-int, rgb, dec, or hex, but can be displayed in any of those, so what would describe the connection between the input and the data storage aside from having 6 types of color selector.

I see this problem already with the combo boxes. Combo boxes store 1 value, but where do we get the options from? keys and/or labels form existing records, some other table or field ?

Is there is way to already create a view model for a table ? Perhaps a single table could have multiple views, raw inputs, or fancy inputs for example. Is there a way to make a default view that maps datatypes to the input controls, and then allow overrides.

**Comment by lokothodida** [comment]
I see your issue with the complexity (and capacity for certain content to take any possible field type), but am very confused with the dataviews explanation.

Are you asking for something that more thoroughly explains the use of each field type, with examples and suitable candidates for each data type? Or for another way (more dynamic, perhaps) of actually viewing an individual table's schema/field properties?

**Comment by tablatronix** [comment]
So say you have what 5 editor types, 4 combo types ?

What happens when we keep adding more?

So for example datatype could be `editor`
then either inline or somewhere else entirely, you pick the dataviewer for that field, which would be something like markdown, bbcode, wiki.

A combo box would have an input for its values

```
[key]
[key]
[key]

or

[key] [value]
[key] [value]
[key] [value]

```

You could even perhaps have several viewers, admins could load up a view with plaintext for all fields etc.

Of course this works into the view and somehow the validation ( do we even have Validation yet? )

Eventually we wind up with something like this.
- schema field Datatypes
- schema viewer ( defaulted but configurable )
- schema mask ( for validation / sanitation ) (defaulted but configurable)

**Comment by tablatronix** [comment]
oh and of course this makes datatypes/views extendable ( make your own somewhere )

**Comment by lokothodida** [comment]
Hmm....I'm beginning to understand. So we separate the data type from the input type in order to reduce the number of fields there are, then extend the fields by having the _viewer_ dictate how the field input is handled?

E.g. with your editor example, another possible example (but for text inputs) might be:
- Type ---> text
- Viewer ---> alphanum, int, email, url, currency, color
- Mask ---> regex

Might make sanitation slightly awkward, but I think its possible.

In terms of custom field validation, what we currently have is HTML5 regex validation - you can input your own regex to limit what can be put in the field.

**Comment by tablatronix** [comment]
Yeah but for common text types multiple datatypes are fine, i mean there are only so many, but that would be how we would do it internally for those types since we store them all as text for the most part.

Probably want to make sure it remains easy to use for the most part, but some of the advanced stuff like a ton of editors that is probably better.

**Comment by tablatronix** [comment]
Yeah sorry, when you have this giant structure in your head its hard to get it out sometimes.
This is just one little piece of the architecture in my head, its really too much to implement but i guess i should at least write it down one day since i have been thinking about it since matrix was created.

**Comment by lokothodida** [comment]
Okay then. I think I understand, and it is viable. It just needs to time/energy to implement. In the long run it'd make adding extra field masks a lot easier.
